linfa-reduction/src/pca.rs

```diff
@@ -0,0 +1,325 @@
+//! Principal Component Analysis
+//!
+//! Principal Component Analysis is a common technique for data and dimensionality reduction. It
+//! reduces the dimensionality of the datawhile retaining most of the variance. This is
+//! done by projecting the data to a lower dimensional space with SVD and eigenvalue analysis. This
+//! implementation uses the `TruncatedSvd` routine in `ndarray-linalg` which employs LOBPCG.
+//!
+//! # Example
+//!
+//! ```
+//! use linfa::traits::Fit;
+//! use linfa_reduction::Pca;
+//!
+//! let dataset = linfa_datasets::iris();
+//!
+//! // apply PCA projection along a line which maximizes the spread of the data
+//! let embedding = Pca::params(1)
+//!     .fit(&dataset);
+//! ```
+//!
+use ndarray::{Array1, Array2, ArrayBase, Axis, Data, Ix2};
+use ndarray_linalg::{TruncatedOrder, TruncatedSvd};
+#[cfg(feature = "serde")]
+use serde_crate::{Deserialize, Serialize};
+
+use linfa::{
+    dataset::{Records, Targets},
+    traits::{Fit, Predict},
+    DatasetBase, Float,
+};
+
+/// Pincipal Component Analysis parameters
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate")
+)]
+pub struct PcaParams {
+    embedding_size: usize,
+    apply_whitening: bool,
+}
+
+impl PcaParams {
+    /// Apply whitening to the embedding vector
+    ///
+    /// Whitening will scale the eigenvalues of the transformation such that the covariance will be
+    /// unit diagonal for the original data.
+    pub fn whiten(mut self, apply: bool) -> Self {
+        self.apply_whitening = apply;
+
+        self
+    }
+}
+
+/// Fit a PCA model given a dataset
+///
+/// The Principal Component Analysis takes the records of a dataset and tries to find the best
+/// fit in a lower dimensional space such that the maximal variance is retained.
+///
+/// # Parameters
+///
+/// * `dataset`: A dataset with records in N dimensions
+///
+/// # Returns
+///
+/// A dataset with records in M < N dimension reduced, such that most variance is retained
+impl<'a, T: Targets> Fit<'a, Array2<f64>, T> for PcaParams {
+    type Object = Pca<f64>;
```

Comment:
yes because the `MagnitudeCorrection` trait is not public in the [API](https://github.com/rust-ndarray/ndarray-linalg/blob/master/ndarray-linalg/src/lobpcg/svd.rs#L176) and is needed to satisfy the trait bound :/

Reply:
I will publish a PR in the `ndarray-linalg` crate but this will need some time till it's upstream

Reply:
👍🏻